pkg/terraform/exec/plugins/plugin_versions_generate.go

```diff
@@ -0,0 +1,249 @@
+// +build tools
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"golang.org/x/mod/modfile"
+	"golang.org/x/mod/module"
+)
+
+type TFPluginInfo struct {
+	Module  module.Version
+	Name    string
+	Tag     string
+	Version string
+}
+
+type TFPlugins struct {
+	Pkg          string
+	TFPluginInfo []TFPluginInfo
+}
+
+func getTemplate() string {
+	return `// Code generated by plugin_versions_generate.go; DO NOT EDIT.
+
+package {{ .Pkg }}
+
+{{ with .TFPluginInfo }}
+        {{- range . }}
+
+func Get{{ .Name }}Version() string {
+	return "{{ .Version }}"
+}
+
+        {{- end }}
+{{- end}}
+`
+}
+
+// getTerraformPluginInfo finds terraform providers used in go.mod, does any
+// necessary replacements and saves into our plugin map.
+func getTerraformPluginInfo(file *modfile.File, plugins *map[string]*TFPluginInfo) {
+
+	// terraform provider and name regex
+	re := regexp.MustCompile(`^[[:alnum:]][[:alnum:]\.\-/]+(terraform-provider-([[:alnum:]]+))`)
+
+	// Find all terraform providers
+	for _, r := range file.Require {
+		matches := re.FindStringSubmatch(r.Mod.Path)
+		if len(matches) == 3 {
```

Comment:
I don't see a compelling reason to check for a length match. The result will either be nil, or all of the submatches will be populated. If the slice is not nil and the length does not match the expected length, then that is a coding error. We should not have code that explicitly and silently skips coding errors.
```suggestion
		if matches != nil {
```

Reply:
Thanks for the tip, and this is good to know. I've updated all instances where I was counting the regular expression matches.